File: backend/routes/metadata.py
# -*- coding: utf-8 -*-
"""元数据API接口（v2 - 查新表）"""
from flask import Blueprint, jsonify
from backend.models_v2 import AggVendorDaily, FactConvContent, DimAccount, AggXhsNote
from sqlalchemy import distinct, func
from datetime import datetime, date
import logging
from backend.utils.decorators import handle_exceptions
from backend.utils.agency_mapper import get_all_shorts, expand_short_to_fulls

bp = Blueprint('metadata', __name__)
logger = logging.getLogger(__name__)


@bp.route('/metadata', methods=['GET'])
@handle_exceptions
def get_metadata():
    from backend.database import db

    try:
        platforms = [r[0] for r in db.session.query(distinct(AggVendorDaily.平台))
                     .filter(AggVendorDaily.平台.isnot(None), AggVendorDaily.平台 != '')
                     .order_by(AggVendorDaily.平台).all()]
    except Exception:
        platforms = ['腾讯', '抖音', '小红书']

    try:
        agencies = get_all_shorts()
    except Exception:
        agencies = []

    try:
        bms = set(r[0] for r in db.session.query(distinct(AggVendorDaily.业务模式))
                  .filter(AggVendorDaily.业务模式.isnot(None), AggVendorDaily.业务模式 != '').all())
        business_models = sorted(bms)
    except Exception:
        business_models = []

    try:
        dr = db.session.query(
            func.min(AggVendorDaily.日期).label('min'),
            func.max(AggVendorDaily.日期).label('max')
        ).first()
        date_range = {'start': str(dr.min) if dr.min else None, 'end': str(dr.max) if dr.max else None}
    except Exception:
        date_range = {'start': None, 'end': None}

    try:
        xdr = db.session.query(
            func.min(AggXhsNote.发布时间).label('min'),
            func.max(AggXhsNote.发布时间).label('max')
        ).first()
        xhs_notes_date_range = {'start': str(xdr.min)[:10] if xdr.min else None, 'end': str(xdr.max)[:10] if xdr.max else None}
    except Exception:
        xhs_notes_date_range = {'start': None, 'end': None}

    return jsonify({
        'success': True,
        'data': {
            'platforms': platforms,
            'agencies': [{'value': a, 'label': a, 'full_names': expand_short_to_fulls([a])} for a in agencies],
            'agency_full_map': {s: expand_short_to_fulls([s]) for s in agencies},
            'business_models': business_models,
            'date_range': date_range,
            'xhs_notes_date_range': xhs_notes_date_range,
        }
    })


def get_data_status():
    from backend.models_v2 import (AggVendorDaily, AggXhsNote, FactConvContent,
                                   FactConvAppmarket, AggDailyChannelOpen, FactAppmarketPlanDaily)
    from backend.database import db
    today = date.today()

    # group 字段供前端 DataFreshness 按分组渲染；分组对齐数据导入页业务域
    #（投放数据 / 线索数据 / 开户数据），只监控带日期的业务事实/聚合表，维度表无日期不纳入。
    sources = [
        {'key': 'vendor_daily',            'model': AggVendorDaily,          'date_field': AggVendorDaily.日期,             'name': '厂商广告投放分析', 'group': 'delivery', 'order': 1},
        {'key': 'fact_appmarket_plan_daily','model': FactAppmarketPlanDaily, 'date_field': FactAppmarketPlanDaily.日期,   'name': '广告计划维度明细', 'group': 'delivery', 'order': 2},
        {'key': 'xhs_note',                'model': AggXhsNote,              'date_field': AggXhsNote.发布时间,             'name': '小红书笔记维度明细', 'group': 'delivery', 'order': 3},
        {'key': 'fact_conv_content',       'model': FactConvContent,         'date_field': FactConvContent.线索日期,        'name': '企微明细',     'group': 'leads',    'order': 1},
        {'key': 'fact_conv_appmarket',     'model': FactConvAppmarket,       'date_field': FactConvAppmarket.下载日期,      'name': 'APP下载明细',  'group': 'leads',    'order': 2},
        {'key': 'agg_daily_channel_open',  'model': AggDailyChannelOpen,     'date_field': AggDailyChannelOpen.时间区间,   'name': '开户渠道分析', 'group': 'open',     'order': 1},
    ]
    results = {}
    for cfg in sources:
        k = cfg['key']
        try:
            latest = db.session.query(func.max(cfg['date_field'])).scalar()
            base = {'name': cfg['name'], 'group': cfg['group'], 'order': cfg['order']}
            if latest:
                latest_str = str(latest)[:10]
                try:
                    d = datetime.strptime(latest_str, '%Y-%m-%d').date()
                    days = (today - d).days
                except Exception:
                    days = 0
                status = 'normal' if days <= 5 else ('warning' if days <= 14 else 'critical')
                results[k] = {**base, 'latest_date': latest_str, 'days_ago': days, 'status': status}
            else:
                results[k] = {**base, 'latest_date': None, 'status': 'no_data'}
        except Exception as e:
            results[k] = {'name': cfg['name'], 'group': cfg['group'], 'order': cfg['order'], 'status': 'error', 'error': str(e)}
    return results


@bp.route('/data-freshness', methods=['GET'])
@handle_exceptions
def get_data_freshness():
    return jsonify({'success': True, 'data': get_data_status()})

File: frontend-react/src/services/mobileHandlers/metadata.ts
/**
 * 移动端本地路由处理器 —— 元数据 / 全局筛选项 (metadata)
 *
 * 由 mobileRouteHandler.ts 按报表域拆分而来，SQL 口径与 Flask 后端保持一致。
 */
import { querySql } from '../mobileSqlite';
import { type Row } from './shared';
// ============================================================================
// 元数据 (metadata) — 全局筛选项数据源
// ============================================================================
// 对齐后端 backend/routes/metadata.py 的 /metadata 路由
// 供 AgencyFilter / PlatformFilter / BusinessModelFilter 三个筛选器使用
// 移动端无 dim_account 表，代理商直接用 agg_vendor_daily.厂商 全称（无简称归一化）

export async function handleMetadata(): Promise<any> {
  // 平台列表（与后端兜底一致：失败时返回 ['腾讯', '抖音', '小红书']）
  let platforms: string[] = [];
  try {
    const rows = await querySql<Row>(
      `SELECT DISTINCT "平台" as v FROM agg_vendor_daily
       WHERE "平台" IS NOT NULL AND "平台" != ''
       ORDER BY "平台"`
    );
    platforms = rows.map(r => String(r.v));
  } catch {
    platforms = ['腾讯', '抖音', '小红书'];
  }

  // 代理商（移动端无 dim_account，直接用 厂商 全称，简称=全称）
  let agency_names: string[] = [];
  try {
    const rows = await querySql<Row>(
      `SELECT DISTINCT "厂商" as v FROM agg_vendor_daily
       WHERE "厂商" IS NOT NULL AND "厂商" != ''
       ORDER BY "厂商"`
    );
    agency_names = rows.map(r => String(r.v));
  } catch {
    agency_names = [];
  }
  const agencies = agency_names.map(a => ({ value: a, label: a, full_names: [a] }));
  const agency_full_map: Record<string, string[]> = {};
  for (const a of agency_names) agency_full_map[a] = [a];

  // 业务模式
  let business_models: string[] = [];
  try {
    const rows = await querySql<Row>(
      `SELECT DISTINCT "业务模式" as v FROM agg_vendor_daily
       WHERE "业务模式" IS NOT NULL AND "业务模式" != ''
       ORDER BY "业务模式"`
    );
    business_models = rows.map(r => String(r.v));
  } catch {
    business_models = [];
  }

  // 日期范围（agg_vendor_daily.日期）
  const date_range: { start: string | null; end: string | null } = { start: null, end: null };
  try {
    const rows = await querySql<Row>(
      `SELECT MIN("日期") as min, MAX("日期") as max FROM agg_vendor_daily`
    );
    const r = rows[0];
    if (r?.min) date_range.start = String(r.min).slice(0, 10);
    if (r?.max) date_range.end = String(r.max).slice(0, 10);
  } catch {
    // 保持 null
  }

  // 小红书笔记日期范围（agg_xhs_note.发布时间，截取前 10 位 YYYY-MM-DD）
  const xhs_notes_date_range: { start: string | null; end: string | null } = { start: null, end: null };
  try {
    const rows = await querySql<Row>(
      `SELECT MIN("发布时间") as min, MAX("发布时间") as max FROM agg_xhs_note`
    );
    const r = rows[0];
    if (r?.min) xhs_notes_date_range.start = String(r.min).slice(0, 10);
    if (r?.max) xhs_notes_date_range.end = String(r.max).slice(0, 10);
  } catch {
    // 保持 null
  }

  return {
    platforms,
    agencies,
    agency_full_map,
    business_models,
    date_range,
    xhs_notes_date_range,
    accounts: [],  // 移动端不提供账号映射列表（features 已禁用相关页面）
  };
}
